src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsBufferAlloc.cs

```diff
@@ -2,37 +2,104 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
-using System.Diagnostics;
+using System.Buffers;
+using System.Threading.Tasks;
 
 namespace Microsoft.Data.SqlClientX.IO
 {
     /// <summary>
     /// Provides buffer management for reader/writer implementations
     /// </summary>
-    internal class TdsBufferAlloc 
+    internal class TdsBufferManager
     {
-        private readonly byte[] _buffer;
+        private readonly ArrayPool<byte> _bufferPool = ArrayPool<byte>.Shared;
+
+        private byte[] RentBuffer(int size) => _bufferPool.Rent(size);
+
+        private void ReturnBuffer(byte[] buffer) => _bufferPool.Return(buffer, clearArray: true);
+
+        #region Public APIs
+
+        /// <summary>
+        /// Operation that consumes rented buffer and executes synchronously, returns result for pass-through.
+        /// </summary>
+        /// <typeparam name="TResult"></typeparam>
+        /// <param name="buffer"></param>
+        /// <returns></returns>
+        public delegate TResult Operation<TResult>(byte[] buffer);
+
+        /// <summary>
+        /// Operation that consumes rented buffer and executes asynchronously.
+        /// </summary>
+        /// <param name="buffer"></param>
+        /// <returns></returns>
+        public delegate ValueTask OperationAsync(byte[] buffer);
 
         /// <summary>
-        /// Default constructor, instantiates buffer with 17 bytes byte array.
+        /// Operation that consumes rented buffer and executes asynchronously.
         /// </summary>
-        public TdsBufferAlloc()
-        { 
-            // 17: max bytes required for storage in buffer for decimal
-            _buffer = new byte[17];
+        /// <param name="buffer"></param>
+        /// <returns></returns>
+        public delegate ValueTask<TResult> OperationAsync<TResult>(byte[] buffer);
+
+        /// <summary>
+        /// Performs <paramref name="operation"/> asynchronously with rented buffer.
+        /// </summary>
+        /// <param name="size"></param>
+        /// <param name="operation"></param>
+        /// <returns></returns>
+        public ValueTask DoWithRentedBuffer(int size, OperationAsync operation)
+        {
+            byte[] buffer = RentBuffer(size);
+            try
+            {
+                return operation(buffer);
+            }
+            finally
+            {
+                ReturnBuffer(buffer);
```

Comment:
Note: in runtime, the general pattern is usually not to return the buffer on exception, but let the GC claim it. There are multiple reasons for that, e.g. no try/finally and after an exception the state might not be well understood/defined  anymore.

Reply:
Simplified back to dedicated buffer, to avoid complications of array pool, also noticed array pool allocates a lot more than requested, so may not be the best solution applying to us.